types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,658 @@
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(lease);
+		return;
+	}
+
+	lease->device = request->device;
+	lease->resource = lease_resource;
+	wl_list_insert(&lease->device->leases,
+			wl_resource_get_link(lease_resource));
+
+	request->lease = lease;
+
+	wl_resource_set_implementation(lease_resource, &lease_impl,
+			lease, drm_lease_v1_handle_resource_destroy);
+
+	if (request->invalid) {
+		/* Pre-emptively reject invalid lease requests */
+		wp_drm_lease_v1_send_finished(lease->resource);
+	} else {
+		wlr_signal_emit_safe(&request->device->manager->events.request,
+				request);
+	}
+
+	/* Request is done */
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_request_v1_interface lease_request_impl = {
+	.request_connector = drm_lease_request_v1_handle_request_connector,
+	.submit = drm_lease_request_v1_handle_submit,
+};
+
+static void drm_lease_device_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static void drm_lease_device_v1_handle_stop(
+		struct wl_client *client, struct wl_resource *resource) {
+	wp_drm_lease_device_v1_send_finished(resource);
+	wl_resource_destroy(resource);
+}
+
+static void drm_lease_device_v1_handle_create_lease_request(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device =
+		drm_lease_device_v1_from_resource(resource);
+	if (!device) {
+		wlr_log(WLR_ERROR, "Failed to create lease request, "
+				"wlr_drm_lease_device_v1 has been destroyed");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	struct wlr_drm_lease_request_v1 *req =
+		calloc(1, sizeof(struct wlr_drm_lease_request_v1));
+	if (!req) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_request_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Created request %p", req);
+
+	uint32_t version = wl_resource_get_version(resource);
+	struct wl_resource *request_resource = wl_resource_create(client,
+		&wp_drm_lease_request_v1_interface, version, id);
+	if (!request_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(req);
+		return;
+	}
+
+	req->device = device;
+	req->resource = request_resource;
+	req->connectors = NULL;
+	req->n_connectors = 0;
+
+	wl_resource_set_implementation(request_resource, &lease_request_impl,
+		req, drm_lease_request_v1_handle_resource_destroy);
+
+	wl_list_insert(&device->requests, wl_resource_get_link(request_resource));
+}
+
+static struct wp_drm_lease_device_v1_interface lease_device_impl = {
+	.stop = drm_lease_device_v1_handle_stop,
+	.create_lease_request = drm_lease_device_v1_handle_create_lease_request,
+};
+
+static void drm_connector_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_resource_set_user_data(resource, NULL);
```

Comment:
This `wl_resource_set_user_data` isn't necessary, since the resource is being destroyed anyways.